File: reconstructitinerary.js
//Objective is to reconstruct a 2-D array of departures and destinations into
//a 1-D array of a list of consecutive cities for a trip. Two of the same
//destinations are sorted lexicographically

let tickets = [["MUC", "LHR"], ["JFK", "MUC"], ["SFO", "SJC"], ["LHR", "SFO"]]


//O(nlogn) solution that sorts all of the departures and destinations into a map.
//Destinations are then sorted lexicographically. Finally, a dfs traversal
//is done to find the order of cities.

let map = {}
let result = []

//Map all departures to their destinations
for (let ticket of tickets) {
    let departure = ticket[0]
    let destination = ticket[1]

    if (map[departure] == undefined) {
        map[departure] = [destination]
    } else {
        map[departure].push(destination)
    }
}

//Sort the destinations lexicographically
for (let departure in map) {
    map[departure].sort()
}

//DFS starting w/ 'JFK', we keep connecting our paths until we reach a place
//with no destination to go to
function dfs(node) {
    while (map[node] && map[node].length) {
        dfs(map[node].shift())
    }
    result.push(node)
}
dfs('JFK')

return result.reverse()

File: reconstructitinerary.java
//Java Solution

class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        Map<String, LinkedList<String>> graph = new HashMap<>();
        
        for (int i = 0; i < tickets.size(); i++) {
            String departure = tickets.get(i).get(0);
            String destination = tickets.get(i).get(1);
            
            if (!graph.containsKey(departure)) {
                graph.put(departure, new LinkedList<String>());
                graph.get(departure).add(destination);
            } else {
                graph.get(departure).add(destination);
            }
        }
        
        for (Map.Entry<String, LinkedList<String>> entry : graph.entrySet()) {
            Collections.sort(entry.getValue());
        }
        
        List<String> result = new ArrayList<>();
        dfs("JFK", graph, result);
        
        List<String> reversed = new ArrayList<>();
        for (int i = result.size() - 1; i >= 0; i--) {
            reversed.add(result.get(i));
        }
        
        return reversed;
    }
    
    public void dfs(String node, Map<String, LinkedList<String>> graph, List<String> result) {
        while (graph.containsKey(node) && graph.get(node).size() > 0) {
            dfs(graph.get(node).removeFirst(), graph, result);
        }
        result.add(node);
    }
}